shared/templates/firefox_policy-setting/bash.template

```diff
@@ -0,0 +1,58 @@
+# platform = Mozilla Firefox
+# strategy = policy
+
+firefox_cfg="distribution/policies.json"
+firefox_dirs="/usr/lib/firefox /usr/lib64/firefox /usr/local/lib/firefox /usr/local/lib64/firefox"
+permissions=644
+
+{{{ find_python() }}}
+# Check the possible Firefox install directories
+for firefox_dir in ${firefox_dirs}; do
+    # If the Firefox directory exists, then Firefox is installed
+    if [ -d "${firefox_dir}" ]; then
+        if ! [ -d "${firefox_dir}/distribution" ]; then
+            mkdir -p "${firefox_dir}/distribution"
+            chmod 755 "${firefox_dir}/distribution"
+        fi
+        # Make sure the Firefox .cfg file exists and has the appropriate permissions
+        if ! [ -f "${firefox_dir}/${firefox_cfg}" ] ; then
+            echo "{" > "${firefox_dir}/${firefox_cfg}"
+            echo "    \"policies\": {" >> "${firefox_dir}/${firefox_cfg}"
+            echo "    }" >> "${firefox_dir}/${firefox_cfg}"
+            echo "}" >> "${firefox_dir}/${firefox_cfg}"
+            chmod ${permissions} "${firefox_dir}/${firefox_cfg}"
+        fi
+        # If the key exists, change it. Otherwise, add it to the config_file.
+        if [ -x ${__REMEDIATE_PYTHON} ]; then
+            echo """
+import json
+_file=open('${firefox_dir}/${firefox_cfg}', 'r')
+_tree=json.load(_file)
+_file.close()
+{{% for policy_item in POLICIES %}}
+{{% for p in policy_item.subpath %}}
+if '{{{ p }}}' in _tree{{{ policy_item.path_python[loop.revindex] }}}:
+   pass
+else:
+   _tree{{{ policy_item.path_python[loop.revindex0] }}} = dict()
+{{% endfor %}}
+_tree{{{ policy_item.path_python[0] }}}['{{{ policy_item.parameter }}}'] = {{{ policy_item.value_escaped }}}
+{{% endfor %}}
+_file=open('${firefox_dir}/${firefox_cfg}', 'w')
+json.dump(_tree, _file, indent=4, sort_keys=True)
+_file.close()
+""" | ${__REMEDIATE_PYTHON}
```

Comment:
You use an interesting approach here that you call Python from a Bash remediation. This is not a blocker, but it would be probably better to have an all-Python remediation format that would get translated into Bash by the build system. Anyway, the Bash way of remediating this would [probably involve](https://stackoverflow.com/questions/24942875/change-json-file-by-bash-script) the `jq` tool - do you have any thoughts about this?

Reply:
I agonized over JSON when doing the initial implementation. If FFOX hadn't put JSON only policy configurations in place, I would have been happy to just implement the new manual STIG items in the preference file.

I didn't have a pure bash implementation available to me that actually let me set fields (and I looked) and jq is not installed "standard". I don't think it is even in the standard repos for RHEL7 (the platform I must support and the point of me making this PR).

There are some existing remediation scripts for Chrome, but those assume all of the attributes are right at the top of the file (the simplest case), which quickly proved to be useless to me.

If someone wants to write/modify a pure shell implementation on their own time (there are a few to start from) that can in-place modify an existing file, I would appreciate it, but I could not justify the cost.

Python is near guaranteed to be installed on Linux across the distros one would also have Firefox on *and* evaluating the DISA STIG. That's why I try to be 2/3 compliant and only use the standard library.

What I *really* want is for OVAL spec gets a native JSON handler and on our side we had better assurances about handling JSON. The OVAL regex I have only really works for these configurations.

IMHO it's a cursed format and I wish that developers would stop using it for configuration files at rest. Or failing that, the standard tooling would get better.

Reply:
As for "Python remediation format that gets translated to Bash by the build system", that's way more complexity than I was willing to bite off for this effort, as the Bash-form of the remediation _still_ has to interpret any existing files and modify with a 100% accuracy rate. Hacking things together with regex is a kludge I reserve for when I have zero other choice (such as for our OVAL checker).

Reply:
Thanks for the comment, it really clarifies that the Python approach is a right pick.